src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -35,6 +35,519 @@ extern "C" {
 using namespace gpdxl;
 using namespace gpmd;
 
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::GroupingListContainsPrimaryKey
+//
+//	@doc:
+//		Checks if list grouping_list contains all columns from conkeys of
+//		Primary Key in relation conrelid. grouping_list is treated as flat
+//		list of SortGroupClauses, other nodes are ignored.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
+											   List *grouping_list,
+											   List *conkeys, Oid conrelid)
+{
+	BOOL result = false;
+	ULONG found_col_cnt = 0;
+	ListCell *lc_conkey;
+
+	// go over every conkey of constraint
+	ForEach(lc_conkey, conkeys)
+	{
+		AttrNumber attnum = lfirst_int(lc_conkey);
+		BOOL found_col = false;
+		ListCell *lgc;
+
+		// search for conkey in grouping list
+		ForEach(lgc, grouping_list)
+		{
+			Node *grpcl = (Node *) lfirst(lgc);
+
+			if (NULL == grpcl)
```

Comment:
May be assert?

Reply:
we can't change to assert, as it can be NULL here when grouping is done by grouping sets with empty set, for example ```group by grouping sets ((a), ())```